internal/route/api/v1/repo/contents.go

```diff
@@ -0,0 +1,209 @@
+	HTML string `json:"html"`
+}
+
+func GetContents(c *context.APIContext) {
+	treeEntry, err := c.Repo.Commit.GetTreeEntryByPath(c.Repo.TreePath)
+	if err != nil {
+		c.NotFoundOrServerError("GetTreeEntryByPath", git.IsErrNotExist, err)
+		return
+	}
+
+	// TODO: figure out the best way to do this
+	// :base-url/:username/:project/raw/:refs/:path
+	templateDownloadURL := "%s/%s/%s/raw/%s"
+	// :base-url/repos/:username/:project/contents/:path
+	templateSelfLink := "%s/repos/%s/%s/contents/%s"
+	// :baseurl/repos/:username/:project/git/trees/:sha
+	templateGitURLLink := "%s/repos/%s/%s/trees/%s"
+	// :baseurl/repos/:username/:project/tree/:sha
+	templateHTMLLLink := "%s/repos/%s/%s/tree/%s"
+
+	gitURL := fmt.Sprintf(templateGitURLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	htmlURL := fmt.Sprintf(templateHTMLLLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), treeEntry.ID.String())
+	selfURL := fmt.Sprintf(templateSelfLink, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath)
+
+	gr := &repoContents{
+		DownloadURL: fmt.Sprintf(templateDownloadURL, c.BaseURL, c.Params(":username"), c.Params(":reponame"), c.Repo.TreePath),
+		Size:        treeEntry.Size(),
+		Name:        treeEntry.Name(),
+		Path:        c.Repo.TreePath,
+		Sha:         treeEntry.ID.String(),
+		Links: Links{
+			Git:  gitURL,
+			Self: selfURL,
+			HTML: htmlURL,
+		},
+		URL:     selfURL,
+		GitURL:  gitURL,
+		HTMLURL: htmlURL,
+	}
+	// A tree entry can only be one of the following types:
+	// 1. Tree (Directory)
+	// 2. SubModule
+	// 3. SymLink
+	// 4. Blob
+	if treeEntry.IsSubModule() {
+		gr.Type = "submodule"
+		parsedURL, err := url.Parse(c.BaseURL)
+		if err != nil {
+			c.ServerError("ErrorURLParse", err)
+		}
+		host := parsedURL.Host
+		submoduleURL := fmt.Sprintf("git://%s/%s/%s.git", host, c.Params(":name"), c.Params(":reponame"))
+		gr.SubmoduleGitURL = submoduleURL
+		c.JSONSuccess(gr)
+		return
+
+	} else if treeEntry.IsLink() {
+		gr.Type = "symlink"
+		gr.Target = c.Repo.TreePath
+		c.JSONSuccess(gr)
+		return
+
+	} else if gr.Type == "blob" { // tree entry is a blob
+		gr.Type = "blob"
+		b, err := getBase64EncodedBlob(c)
+
+		if err != nil {
+			c.ServerError("GetBlobContent", err)
+			return
+		}
+
+		gr.Content = b
+		c.JSONSuccess(gr)
+		return
+	} else { // treeEntry is a directory
+		dirTree, err := c.Repo.GitRepo.GetTree(treeEntry.ID.String())
+		if err != nil {
+			c.NotFoundOrServerError("GetGitDirTree", git.IsErrNotExist, err)
+			return
+		}
+
+		entries, err := dirTree.ListEntries()
+		if err != nil {
+			c.NotFoundOrServerError("ListDirTreeEntries", git.IsErrNotExist, err)
+			return
+		}
+
+		results, err := AppendDirTreeEntries(entries, c)
+
+		if err != nil {
+			c.NotFoundOrServerError("AppendDirTreeEntries", git.IsErrNotExist, err)
+			return
+
+		}
+		c.JSONSuccess(results)
+		return
+	}
+}
+
+func getBase64EncodedBlob(c *context.APIContext) (string, error) {
```

Comment:
I think having this function independently is not worth it as of this PR.

Should merge back to the `GetContents` function.

Reply:
I initially wanted to have it as the same function. But the function grew really big and since smaller funcs usually make things a tad bit clearer, I just extracted it into it's own func. 

I'm tempted to ask if we can have this as a separate func, but sure, I can follow your suggestion if you want me to. Please do let me know.

Reply:
TBH this function is abstracted very well... (at least it should only accept a blob object, which is what it is asking for, not the whole context) Let's put it back first.

Reply:
Sure, will do that.

Reply:
Fixed.